vars/buildNodeJs.groovy

```diff
@@ -0,0 +1,95 @@
+def call(Map config=[:], Closure body={}) {
+  def containerSrcFolder = '\\/home\\/node'
+  def localSrcFolder = '.'
+  def lcovFile = './test-output/lcov.info'
+  def sonarQubeEnv = 'SonarQube'
+  def sonarScanner = 'SonarScanner'
+  def qualityGateTimeout = 10
+  def repoName = ''
+  def pr = ''
+  def containerTag = ''
+  def mergedPrNo = ''
+
+  node {
+    checkout scm
+    try {
+      stage('Start') {
+        echo "pipeline started"
+      }
+      stage('Set GitHub status as pending') {
+        build.setGithubStatusPending()
+      }
+      stage('Set PR, and containerTag variables') {
+        (repoName, pr, containerTag, mergedPrNo) = build.getVariables(version.getPackageJsonVersion())
+      }
+      stage('Helm lint') {
+        test.lintHelm(repoName)
+      }
+      stage('Build test image') {
+        build.buildTestImage(DOCKER_REGISTRY_CREDENTIALS_ID, DOCKER_REGISTRY, repoName, BUILD_NUMBER)
+      }
+      stage('Run tests') {
+        build.runTests(repoName, repoName, BUILD_NUMBER)
+      }
+      stage('Create JUnit report') {
+        test.createReportJUnit()
+      }
+      stage('Fix lcov report') {
+        utils.replaceInFile(containerSrcFolder, localSrcFolder, lcovFile)
+      }
+      stage('SonarQube analysis') {
+        test.analyseCode(sonarQubeEnv, sonarScanner, test.buildCodeAnalysisDefaultParams(repoName))
+      }
+      stage("Code quality gate") {
+        test.waitForQualityGateResult(qualityGateTimeout)
+      }
+      stage('Push container image') {
+        build.buildAndPushContainerImage(DOCKER_REGISTRY_CREDENTIALS_ID, DOCKER_REGISTRY, repoName, containerTag)
+      }
+      if (pr != '') {
+        stage('Verify version incremented') {
+          version.verifyPackageJsonIncremented()
+        }
+        stage('Helm install') {
+          helm.deployChart(config.environment, DOCKER_REGISTRY, repoName, containerTag)
+          echo "Build available for review at https://ffc-demo-$containerTag.$INGRESS_SERVER"
+        }
+      }
+      if (pr == '') {
+        stage('Publish chart') {
+          helm.publishChart(DOCKER_REGISTRY, repoName, containerTag)
+        }
+        stage('Trigger GitHub release') {
+          withCredentials([
+            string(credentialsId: 'github-auth-token', variable: 'gitToken')
+          ]) {
+            release.trigger(containerTag, repoName, containerTag, gitToken)
+          }
+        }
+        stage('Trigger Deployment') {
+          withCredentials([
+            string(credentialsId: 'web-deploy-job-name', variable: 'deployJobName'),
+            string(credentialsId: 'web-deploy-token', variable: 'jenkinsToken')
```

Comment:
This function won't work on anything other than `ffc-demo-web`.

This is because `web-deploy-job-name` and `web-deploy-token` are values specific to that repo.  If another pipeline runs this what will happen is the web deployment job will run for every service with a version parameter that may or may not exist.

I think the best way to tackle this is not to use a credential for the job name and instead rely on the convention we already have.  Ie we have agreed that all deployment jobs have the convention `workstream-service-deploy` eg `ffc-demo-web-deploy`

Then the `deploy.trigger` function can be updated to not accept a `deployJobName` but instead the `repoName` value we already have and before running the command it can append `-deploy` to the end.

That just leaves the issue of the token value.  We can resolve that by following the same convention of using the repo name like we do for everything else.  Instead of `web-deploy-token` the credential should be called `ffc-demo-web-deploy-token`  then in the credentialId in this function we can just have 

```
withCredentials([
  string(credentialsId: "$repoName-deploy-token", variable: 'jenkinsToken')
  {
  ...
```

We should add new credentials rather than renaming the existing ones because a) you can't rename a credential Id and b) it won't disrupt the services before we switch them over.

These new conventions should feed into what teams need to be doing to use the `buildNodeJs` prerequisites mentioned in the other comment

Reply:
Thanks for this John, great spot.